package com.team404.synco.virtualmeeting.service;

import com.team404.synco.virtualmeeting.entity.Room;
import com.team404.synco.virtualmeeting.entity.Recording;
import com.team404.synco.virtualmeeting.entity.RoomParticipant;
import com.team404.synco.virtualmeeting.repository.RecordingRepository;
import com.team404.synco.virtualmeeting.repository.RoomParticipantRepository;
import com.team404.synco.virtualmeeting.repository.RoomRepository;
import io.livekit.server.*;
import jakarta.persistence.EntityNotFoundException;
import livekit.LivekitEgress.*;
import livekit.LivekitWebhook.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class LiveKitService {

    private final RoomRepository roomRepository;
    private final RecordingRepository recordingRepository;
    private final RoomParticipantRepository roomParticipantRepository;

    public void handleWebhook(WebhookEvent event) {
        switch (event.getEvent()) {
            case "room_started" -> handleRoomStarted(event);
            case "room_finished" -> handleRoomFinished(event);
            case "participant_joined" -> handleParticipantJoined(event);
            case "participant_left" -> handleParticipantLeft(event);
            case "participant_connection_aborted" -> handleParticipantConnectionAborted(event);
            case "track_published" -> handleTrackPublished(event);
            case "track_unpublished" -> handleTrackUnpublished(event);
            case "egress_started" -> handleEgressStarted(event);
            case "egress_updated" -> handleEgressUpdated(event);
            case "egress_ended" -> handleEgressEnded(event);
            case "ingress_started" -> handleIngressStarted(event);
            case "ingress_ended" -> handleIngressEnded(event);
        }
    }

    private void handleRoomStarted(WebhookEvent event) {
        log.info("LiveKit-Webhook(room_started) - {}", event.getRoom());
    }


    private void handleRoomFinished(WebhookEvent event) {
        log.info("LiveKit_Webhook(room_finished) - {}", event.getRoom());

        Room room = roomRepository.findById(Long.valueOf(event.getRoom().getName())).orElseThrow(() -> new EntityNotFoundException("회의를 찾을 수 없습니다."));
        room.endRoom();
    }

    private void handleParticipantJoined(WebhookEvent event) {
        log.info("LiveKit_Webhook(participant_joined) - {}", event.getRoom());
        log.info("LiveKit_Webhook(participant_joined) - {}", event.getParticipant());
    }

    private void handleParticipantLeft(WebhookEvent event) {
        log.info("LiveKit_Webhook(participant_left) - {}", event.getRoom());
        log.info("LiveKit_Webhook(participant_left) - {}", event.getParticipant());
        RoomParticipant participant = roomParticipantRepository.findById(Long.valueOf(event.getParticipant().getIdentity())).orElseThrow(() -> new EntityNotFoundException("없는 화상회의 참가자 입니다."));
        participant.leaveRoom();
    }

    private void handleParticipantConnectionAborted(WebhookEvent event) {
        log.info("LiveKit_Webhook(participant_connection_aborted) - {}", event.getRoom());
        log.info("LiveKit_Webhook(participant_connection_aborted) - {}", event.getParticipant());

        RoomParticipant participant = roomParticipantRepository.findById(Long.valueOf(event.getParticipant().getIdentity())).orElseThrow(() -> new EntityNotFoundException("없는 화상회의 참가자 입니다."));
        participant.leaveRoom();
    }

    private void handleTrackPublished(WebhookEvent event) {
        log.info("LiveKit_Webhook(track_published) - {}", event.getRoom());
        log.info("LiveKit_Webhook(track_published) - {}", event.getParticipant());
        log.info("LiveKit_Webhook(track_published) - {}", event.getTrack());
    }

    private void handleTrackUnpublished(WebhookEvent event) {
        log.info("LiveKit_Webhook(track_unpublished) - {}", event.getRoom());
        log.info("LiveKit_Webhook(track_unpublished) - {}", event.getParticipant());
        log.info("LiveKit_Webhook(track_unpublished) - {}", event.getTrack());
    }

    private void handleEgressStarted(WebhookEvent event) {
        log.info("LiveKit_Webhook(egress_started) - {}", event.getEgressInfo());
    }

    private void handleEgressUpdated(WebhookEvent event) {
        log.info("LiveKit_Webhook(egress_updated) - {}", event.getEgressInfo());
    }

    private void handleEgressEnded(WebhookEvent event) {
        log.info("LiveKit_Webhook(egress_ended) - {}", event.getEgressInfo());
        if (event.getEgressInfo().getStatus() != EgressStatus.EGRESS_COMPLETE) return;

        FileInfo fileInfo = event.getEgressInfo().getFileResults(0);

        Room room = roomRepository.findById(Long.valueOf(event.getEgressInfo().getRoomName())).orElseThrow(() -> new EntityNotFoundException("없는 화상회의 입니다."));
        Recording recording = Recording.fromFileInfo(fileInfo,room);
        recordingRepository.save(recording);
    }

    private void handleIngressStarted(WebhookEvent event) {
        log.info("LiveKit_Webhook(ingress_started) - {}", event.getIngressInfo());
    }

    private void handleIngressEnded(WebhookEvent event) {
        log.info("LiveKit_Webhook(ingress_ended) - {}", event.getIngressInfo());
    }
}